### DISCUSSION TITLE
How to omit certain relation members with flex output?

### DISCUSSION THREAD

**Discussion by cdauth** [osm2pgsql-dev/osm2pgsql#2204] [Q&A]
I am completely new to osm2pgsql and I'm trying to bring the [Openptmap](https://github.com/giggls/openptmap) scripts up to date.

The way that it works right now is that it uses [Osmfilter](https://wiki.openstreetmap.org/wiki/Osmfilter) to keep only certain tags and only objects with certain tags _before_ importing the planet/region file with osm2pgsql. I am trying to rewrite this into an osm2pgsql flex output Lua script so that I can apply deltas to the database later. One thing that I'm struggling to implement is that the [osmfilter file](https://wiki.openstreetmap.org/wiki/Openptmap/Installation#Filter_the_Planet_File) removes any objects with a `railway=platform` tag. Because the removal happens before the import to `osm2pgsql`, such objects will also be omitted from the geometries of relations that contain objects with such tags, because the members simply cannot be found in the imported data.

I'm struggling to implement this in my Lua script. I successfully manage to ignore ways with `railway=platform` in my `osm2pgsql.process_way` script, but if those ways are a member of any relation, they still end up in the geometry of that relation.

Right now I’m trying to reproduce the layout of the previous pgsql output with 3 tables, one for points, one for lines and one for polygons. From what I read [online](https://gis.stackexchange.com/questions/420135/filtering-out-parts-of-a-relation-on-osm2pgsql-import), one way to do it would be to instead create tables for nodes, ways and relations and make the relations table not contain any actual geometries, but make it reference the nodes/ways and pull them in using joins in mapnik.xml. I'm imagining I would have to create a `nodes`, `ways`, `routes`, `routes_members`, `polygons` and `polygons_members` table, where `routes` and `polygons` would contain the relations with their tags and `routes_members` and `polygons_members` tables would reference the relation members. However, I’m not sure what the SQL query to construct the actual relation geometries would actually look like. Specifically, I'm wondering about:
* How to support relations that are members of other relations? The tags of a relation should be applied to all its descendent ways in the hierarchy of member relations. Is this even possible with an SQL join?
* How to construct a multipolygon from the members of a polygon relation?

Is there maybe some example code somewhere that already does this?

**Comment by joto** [comment]
Lets answer the specific questions first:

* Building recursive queries in SQL is painful, but possible with `WITH RECURSIVE` common table expressions (CTE). But you probably don't want to do that. Just implementing limited levels of recursion should be okay with OSM data, I would think.
* You don't want to implement multipolygon construction from OSM relations in SQL. That is done really well and fast by osm2pgsql (well, actually in libosmium which is used by osm2pgsql) which takes into account bad data and knows about all the corner cases. So I'd recommend that for multipolygon relation assembly you use osm2pgsql processing.

Now to the wider question of how to approach processing OSM public transport data with osm2pgsql: The support for special types of relations in osm2pgsql is not great. `type=multipolygon` is really well supported, but other types are not. You can create a (multi)linestring from a route relation, but as you noticed, if that route relation contains platforms, that doesn't work correctly. But merging linestrings into longer linestrings in SQL is reasonably easy (easier than creating multipolygons), so you can go with the approach you suggest, importing members of the pt relations into one table and joining the routes together in the database. There are multiple ways to do the details, but you'd probably need to use two-stage processing. See the [manual](https://osm2pgsql.org/doc/manual.html#stages) and [this architecture description](https://osm2pgsql.org/contribute/how-osm2pgsql-processing-works.html) on that and have a look at the `flex-config/route-relations.lua` example config file in the osm2pgsql repo.

What it boils down to is that in the first stage you identify all `type=route` relations with `route=bus` or whatever that you want to import and mark their member ways. Marked ways are then (re)processed in the second stage where you check the tags to ignore the members that are platforms and only add the ways to some table that contain the actual parts of the route. Then in SQL it becomes a simple operation of GROUPing by relation id and using ST_LineMerge to assemble the whole route. I am not totally up to date with all the intricacies of the PT data model, so it is probably more complex in reality.

I am open to suggestions for adding functionality to osm2pgsql to make all of this easier. But the architecture of osm2pgsql, which essentially is built around the idea that you only look at one OSM object at a time, limits what can be done. Two-stage processing lifts this limitation a little bit, but it is quite difficult to use. Maybe this is something which could be added to the `osm2pgsql-gen` generalization framework which is built around the idea that it can read data after the initial import from SQL tables and processes it again and writes the results back into the database.

**Reply by cdauth** [reply]
Thank you for your detailed response. I now finally have the time to look at this again, and what you have written has been very helpful so far.

It seems like I will be able to find a way to achieve this by keeping things the way they are for multipolygon relations, but resolving the geometries of route relations in the second stage only, although this will be a bit of work to implement.

I am not deep enough in the internals of how osm2pgsql actually works to make a qualified suggestion how it could be improved to support my use case. From my novice point of view it seems like the `object:as_multilinestring()` relation method would be the usual approach to achieve what I am trying to do, and it does seem convenient. The specific functionality that I’m missing for my use case is to exclude relation members based on their _tags_. It seems to me like this is something that could be added somehow, since I supposed the method needs to load the individual member objects anyways to resolve their geometries, so it wouldn’t cause an additional performance regression? Interestingly, even if I ignore a way in the `osm2pgsql.process_way(object)` function both in the first and second stage, its geometry is still added to the relation. However, I suspect that I might be able to exclude members of the relation based on their _role_ by simply deleting them from the `members` property before calling `object:as_multilinestring()`. I will have to try this out, and if it works this could be a feasible solution that wouldn’t even require two-stage processing, as railway platforms can be identified both by their member role as well as by their object tags.

**Reply by joto** [reply]
Deleting objects from `members` has no influence on anything in osm2pgsql, not the multilinestring creation nor anything else. Basically you only get a copy of the data to mess with in Lua that doesn't affect anything else.

I agree it would be nice if osm2pgsql would allow some more flexibility in regards to processing relations and their members. There are some words about that in our [project ideas list](https://osm2pgsql.org/contribute/project-ideas.html). But so far it is unclear how exactly this is going to work.

**Reply by cdauth** [reply]
It took me a while to find the time to look into this again. I have now set up a basic Lua script and Mapnik project to achieve this in the way you have suggested. Right now it seems to me that this is not a feasible approach for most setups.

The problem is that when I import a way, I do not know yet whether it is a member of a route relation (or it will ever be in the future, when using replication). This means that I have to import all the ways (or at least all highways) of the planet with their geometries into my way table. This does not only consumes a huge amount of disk space and import time, but when using replication (which is the main reason why I want to use the Flex Output in the first place), the `planet_osm_ways` in combination with the flat nodes file already indirectly contain all the ways of the planet, so I have to store the whole data twice.

An alternative way to achieve this could be to not store the ways, but construct the route geometry in Mapnik by querying the relation members directly from the `planet_osm_ways` table. I don’t know if this would be even possible with an SQL query, but even if it is, it would most certainly mean that I cannot use a flat nodes file, but the nodes would need to be stored in Postgres. This would itself mean a large additional amount of disk space and a huge additional amount of import time, which again does not seem feasible on a small server.

The approach could be suitable on servers that host a complete map of the planet that displays most features, for example the OSM Carto style. Such a server would have the geometry of all highways in the world in its database anyways, and those could be reused for my public transportation map. However, most small projects do not have the resources to host such a complete map of the planet, so it would be good to have a way to achieve this on a small server as well.

It seems to me that the only good way to achieve this would be if osm2pgsql added an option to support filtering relation members based on their tags. From what I understand, osm2pgsql persists all relations and ways with all their members and tags, and all nodes with their coordinates, to allow reprocessing their geometries during replication. I suppose when a relation is reprocessed and the Lua script calls `:as_multipolygon()` or `:as_multilinestring()` on it, osm2pgsql will run an SQL query to fetch all members and then look up the coordinates of the contained nodes in the flat nodes file. I suspect that it should be possible during the query to retrieve or even already filter the members based on their tags, since they are available as well in the `planet_osm_rels` and `planet_osm_ways` tables. I understand that retrieving the tags would have a performance impact, so it should only happen if the user requests it.

I assume that fetching the member tags and then passing them back to the Lua script for further processing would impact performance too much. So I’m wondering whether it would be an option to pass something like a filter property to `:as_multilinestring()` and the other methods that would tell osm2pgsql to exclude certain members based on their tags. I imagine something like this:
```lua
    geom = object:as_multilinestring({
        exclude_members: { type = 'way', key = 'railway', value = 'platform' }
    })
```

The exact shape of the `exclude_members` property would need some thought in order to allow _and_, _or_ and _not_ logic. The tag filter would probably only work for member relations and ways, and not member nodes, but it could at least allow to exclude member nodes in general.

Do you think this is an approach that could be implemented in osm2pgsql?

**Reply by joto** [reply]
Giving the user access to members of relations from the relation in some way is something we have been thinking about. But it is unclear which form this would take to make it fast and useable for many cases. Your solution with the `exclude_members` seems too restrictive for me.

If  you want a solution that works now, you have to use two-stage processing. With that it is possible to import only the ways into the database that are in certain relations. You still have to run `ST_LineMerge()` or something like it in the database to do the actual merging of the ways into the multilinestring, because you can't use that functionality in Lua.

**Reply by mboeringa** [reply]
@joto 

I think I have a question that relates to the issue @cdauth is struggling with.

What I attempted to do is to build a table with geometries from waterway relations in OpenStreetMap.

Now I use the **object.as_geometrycollection()** function to generate a geometry collection object of e.g. the Alpine Rhine waterway relation (https://www.openstreetmap.org/relation/21222156), and subsequently insert each of its member ways into a table using either **geom:geometries()** or **geom.geometry_n()** and looping over the members.

According to the documentation, object.as_geometrycollection()

>  It will contain all geometries which can be generated from node and way members of the relation in **order of those members**.

will return the members in the order as defined in the relation, if I understand well.

However, if I e.g. want to know the osm ID of the n-th member way, I cannot use the **object.members**. I would have expected **object.members** to return the members of the relation _**in order of those members**_ as well. However, if I attempt to grab e.g. the n-th geometry of the geometry collection using **geom.geometry_n(n)** and also try to grab the n-th member as returned by **object.members** by using **object.members[n]** and subsequently retrieve the osm ID by using the **ref** field of the member table, I get back an osm ID that does not correspond with the n-th way member of the relation, but another osm ID. So it seems that the geometry collection, and the **object.members** Lua table are out of sync, and **object.members** not sorted by the order in which the relation members are stored in the relation.

Note that the relations I noticed this on, like the Alpine Rhine, do not contain any sub-relations, just way members, and no nodes. So I would expect it would be possible to have the geometry collection's members and **object.members** to be in sync.

Of course, it would be nice if I could somehow grab the geometry from the Lua member table, similar to type,ref,role, e.g. **member_table.geometry**, similar to **member_table.type/ref/role**, so that it would be guaranteed to be in sync.

**Reply by mboeringa** [reply]
I now realized that the unexpected osm ID way I see being returned, lies even outside the extent of the Geofabrik "Liechtenstein" extract that I used as input for osm2pgsql and testing of the import.

It is not even in the planet_osm_line table. It _**is**_ a way member of the Rhein relation, which is the second waterway relation next to the Alpine Rhine one that is in the Liechtenstein extract. 

This likely explains the discrepancy, I now realize: the Liechtenstein pbf extract contains all the information of the members of the relation, but not the actual geometries of all members.

Maybe we need an **object.members_with_geometry** iterator, that only returns a table for complete members including full geometry, those that are being included in the output of **object.as_geometrycollection()**?

**Reply by mboeringa** [reply]
Or a **member_table.complete** boolean field added next to **member_table.type/ref/role** that can tell if the member's geometry is available.

Although likely the most robust, as not relying on order, would still be a **member_table.geometry** property that could return a NULL geometry that could be tested with **geometry.isnull()** whenever a member's geometry is not available due to incomplete relation due to extracts.

**Reply by joto** [reply]
As you noticed relations and their members are quite a complex issue. That's why there isn't a better solution for these kinds of problems yet. It is just difficult to define and implement a good (and performant!) solution that takes all cases into account.

**Reply by mboeringa** [reply]
I agree about the complexity, but as suggested a **member_table.geometry** property that returns a NULL geometry whenever a member's geometry cannot be build and that could be tested with **geometry.isnull()** to allow deciding what to do next with it (skip insertion in a table), would I think be really useful. That would at least allow the partial insertion of incomplete relations by inserting those members that _are_ available including their corresponding osm ID as already available via **member_table.ref**.

**Reply by mboeringa** [reply]
In my specific use case, I am not trying to build a single representative geometry for the entire relation, but just grab the individual member ways and insert them as separate ways, while storing a reference to the corresponding relation ID and way ID.

**Reply by joto** [reply]
That doesn't work because
a) there isn't one geometry for a member, there are lots of possible geometries you can create from a member (if the member is a closed way for instance, a linestring or a polygon)
b) for a property we'd need to create the geometry even if it isn't used which creates a lot of overhead

We'll figure something out eventually, but for the moment: this is harder than it looks which is partly due to the OSM data model but partly also due to the way osm2pgsql handles these things internally.
